File: api/src/main/java/com/dumbdogdiner/stickycommands/api/economy/Market.java
/*
 * Copyright (c) 2021 DumbDogDiner <dumbdogdiner.com>. All rights reserved.
 * Licensed under the MIT license, see LICENSE for more information.
 */
package com.dumbdogdiner.stickycommands.api.economy;

import java.util.List;
import lombok.Getter;
import org.bukkit.Material;
import org.bukkit.OfflinePlayer;
import org.jetbrains.annotations.NotNull;

public interface Market {
    @Getter
    public final int PAGE_MAX = 10;

    public Integer latestId();

    /**
     * Get the number of listings on the market
     * @return The number of listings as a {@link Long}
     */
    @NotNull
    public Long getListingCount();

    /**
     * Get all current listings
     * @return {@link List<Listing>} of all listings
     */
    @NotNull
    public List<Listing> getListings(
        @NotNull Listing.SortBy sortBy,
        @NotNull Integer page,
        @NotNull Integer pageSize
    );

    /**
     * Get a list of listings
     * @param sortBy
     * @return {@link List<Listing>} of listings
     */
    @NotNull
    default List<Listing> getListings(@NotNull Listing.SortBy sortBy) {
        return getListings(sortBy, 1, PAGE_MAX);
    }

    /**
     * Get a list of listings
     * @return {@link List<Listing>} of listings
     */
    @NotNull
    default List<Listing> getListings() {
        return getListings(Listing.SortBy.DATE_ASCENDING, 1, PAGE_MAX);
    }

    /**
     * Get a list of listings of a specific material
     * @param material of the listing
     * @param sortBy
     * @param page to get
     * @param pageSize per page
     * @return {@link List<Listing>} of listings
     */
    @NotNull
    public List<Listing> getListingsOfType(
        @NotNull Material material,
        @NotNull Listing.SortBy sortBy,
        @NotNull Integer page,
        @NotNull Integer pageSize
    );

    /**
     * Get a list of listings of a specific material
     * @param material of the listing
     * @return {@link List<Listing>} of listings
     */
    @NotNull
    default List<Listing> getListingsOfType(
        @NotNull Material material,
        @NotNull Integer page
    ) {
        return getListingsOfType(
            material,
            Listing.SortBy.DATE_ASCENDING,
            page,
            PAGE_MAX
        );
    }

    /**
     * Get a list of listings of a specific material
     * @param material of the listing
     * @return {@link List<Listing>} of listings
     */
    @NotNull
    default List<Listing> getListingsOfType(@NotNull Material material) {
        return getListingsOfType(
            material,
            Listing.SortBy.DATE_ASCENDING,
            1,
            PAGE_MAX
        );
    }

    /**
     * Get the listings that a player has listed
     * @param player to get listings of
     * @param sortBy
     * @param page to get
     * @param pageSize per page
     * @return {@link List<Listing>} of listings belonging to a player
     */
    public List<Listing> getListingsOfPlayer(
        @NotNull OfflinePlayer player,
        @NotNull Listing.SortBy sortBy,
        @NotNull Integer page,
        @NotNull Integer pageSize
    );

    /**
     * Get the listings that a player has listed
     * @param player to get listings of
     * @param page to get
     * @return {@link List<Listing>} of listings belonging to a player
     */
    @NotNull
    default List<Listing> getListingsOfOfflinePlayer(
        @NotNull OfflinePlayer player,
        @NotNull Integer page
    ) {
        return getListingsOfPlayer(
            player,
            Listing.SortBy.DATE_ASCENDING,
            page,
            PAGE_MAX
        );
    }

    /**
     * Get the listings that a player has listed
     * @param player to get listings of
     * @return {@link List<Listing>} of listings belonging to a player
     */
    @NotNull
    default List<Listing> getListingsOfOfflinePlayer(
        @NotNull OfflinePlayer player
    ) {
        return getListingsOfPlayer(
            player,
            Listing.SortBy.DATE_ASCENDING,
            1,
            PAGE_MAX
        );
    }

    /**
     * Add a listing to the market
     * @param listing to list
     */
    public void add(@NotNull Listing listing);

    /**
     * Remove a listing from the market
     */
    public void remove(@NotNull Listing listing);
}

File: src/main/kotlin/economy/Market.kt
/*
 * Copyright (c) 2021 DumbDogDiner <dumbdogdiner.com>. All rights reserved.
 * Licensed under the MIT license, see LICENSE for more information.
 */
package com.dumbdogdiner.stickycommands.economy

import com.dumbdogdiner.stickycommands.StickyCommandsKt
import com.dumbdogdiner.stickycommands.aatempmovemeplz.Listing

import com.dumbdogdiner.stickycommands.database.tables.Listings
import org.bukkit.Material
import org.bukkit.OfflinePlayer
import org.jetbrains.exposed.sql.Query
import org.jetbrains.exposed.sql.deleteWhere
import org.jetbrains.exposed.sql.select
import org.jetbrains.exposed.sql.selectAll
import org.jetbrains.exposed.sql.transactions.transaction

// TODO implement listing cache?? (is this actually necessary because I don't think so
class Market : StickyCommandsKt {
    // I truely, deeply lothe companion objects but kotlin is exceedingly stupid
    // and won't let you do constants in classes. At least it doesnt make it too dumb
    // to use in java
    companion object {
        /**
         * Maximum length of a page
         */
        const val PAGE_MAX_SIZE: Int = 10
    }

    private fun query(query: Query, sortBy: Listing.SortBy, page: Int, pageSize: Int): List<Listing> {
        return databaseHandler.getListings(query, sortBy, page, pageSize)
    }

    /**
     * Get the number of listings on the market
     * @return The number of listings as a [Long]
     */
    fun getListings(sortBy: Listing.SortBy, page: Int = 1, pageSize: Int = PAGE_MAX_SIZE): List<Listing> {
        return query(Listings.selectAll(), sortBy, page, pageSize)
    }

    fun getListingsOfType(material: Material, sortBy: Listing.SortBy, page: Int = 1, pageSize: Int = PAGE_MAX_SIZE): List<Listing> {
        return query(Listings.select { (Listings.item eq material.toString()) }, sortBy, page, pageSize)
    }

    /**
     * Get the listings that a player has listed
     *
     * @param player   to get listings of
     * @param sortBy
     * @param page     to get
     * @param pageSize per page
     *
     * @return A list of listings belonging to a player
     */
    fun getListingsOfPlayer(
        player: OfflinePlayer,
        sortBy: Listing.SortBy = Listing.SortBy.DATE_ASCENDING,
        page: Int = 1,
        pageSize: Int = PAGE_MAX_SIZE
    ): List<Listing> {
        return query(Listings.select { (Listings.seller eq player.uniqueId.toString()) }, sortBy, page, pageSize)
    }

    /**
     * @return Gets the latest ID??
     */
    fun latestId(): Int {
        val selectionResult = transaction(databaseHandler.db) {
            Listings.selectAll().firstOrNull()
        }
        return if (selectionResult?.getOrNull(Listings.id) == null) 1 else selectionResult[Listings.id]
    }

    /**
     * Get the number of listings on the market
     *
     * @return The number of listings as a [Long]
     */
    fun getListingCount(): Long {
        var count = 0L
        transaction(databaseHandler.db) {
            count = Listings.selectAll().count()
        }
        return count
    }

    /**
     * Add a listing to the market
     *
     * @param listing to list
     */
    fun add(listing: Listing) {
        databaseHandler.addListing(listing)
    }

    /**
     * Remove a listing from the market
     */
    fun remove(listing: Listing) {
        transaction(databaseHandler.db) {
            Listings.deleteWhere { (Listings.id eq listing.id) }
        }
    }
}
